File: v2/221.py



class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        # 6 star, no idea
        # dp[i][j]表示以坐标（i,j)为右下角的全是1的正方形的最大边长，
        # 状态转移方程 dp[j][i] = min(dp[j-1][i], dp[j][i-1], dp[j-1][i-1]) + 1
        if not matrix:
            return 0
        row_num = len(matrix)
        col_num = len(matrix[0])
        dp = [[0 for _ in range(col_num)] for _ in range(row_num)]
        max_width = 0
        for i in range(col_num):
            for j in range(row_num):
                dp[j][i] = int(matrix[j][i])
                if dp[j][i] and j > 0 and i > 0:
                    dp[j][i] = min(dp[j-1][i], dp[j][i-1], dp[j-1][i-1]) + 1
                max_width = max(max_width, dp[j][i])
        return max_width * max_width

File: go/221.go

//package main


// 6 star, 动态规划，
func maximalSquare(matrix [][]byte) int {
	cols := len(matrix)
	if cols == 0 {return 0}
	rows := len(matrix[0])

	dp := make([][]int, cols)
	var width, temp int
	for index, _ := range dp {
		dp[index] = make([]int, rows)
	}

	for i:=0; i<cols; i++ {
		for j:=0; j < rows; j++ {
			if i == 0 || j == 0 {
				if matrix[i][j] == '0' {
					dp[i][j] = 0
				} else {
					dp[i][j] = 1
					if dp[i][j] > width {width = dp[i][j]}
				}
			} else if matrix[i][j] == '0' {
				dp[i][j] = 0
			} else {
				temp = min([]int{dp[i-1][j], dp[i-1][j-1], dp[i][j-1]})
				dp[i][j] = temp + 1
				if dp[i][j] > width {width = dp[i][j]}
			}

		}
	}
	return width * width

}


func min(nums []int) int {
	minVal := nums[0]
	for _, v:= range nums {
		if v < minVal {
			minVal = v
		}
	}
	return minVal
}


